File: src/cascadia/TerminalApp/fzf/fzf.cpp
#include "pch.h"
#include "fzf.h"

#undef CharLower
#undef CharUpper

using namespace fzf::matcher;

constexpr int16_t ScoreMatch = 16;
constexpr int16_t ScoreGapStart = -3;
constexpr int16_t ScoreGapExtension = -1;
constexpr int16_t BoundaryBonus = ScoreMatch / 2;
constexpr int16_t NonWordBonus = ScoreMatch / 2;
constexpr int16_t CamelCaseBonus = BoundaryBonus + ScoreGapExtension;
constexpr int16_t BonusConsecutive = -(ScoreGapStart + ScoreGapExtension);
constexpr int16_t BonusFirstCharMultiplier = 2;
constexpr size_t npos = std::numeric_limits<size_t>::max();

enum class CharClass : uint8_t
{
    NonWord = 0,
    CharLower = 1,
    CharUpper = 2,
    Digit = 3,
};

static std::vector<UChar32> utf16ToUtf32(std::wstring_view text)
{
    const UChar* data = reinterpret_cast<const UChar*>(text.data());
    int32_t dataLen = static_cast<int32_t>(text.size());
    int32_t cpCount = u_countChar32(data, dataLen);

    std::vector<UChar32> out(cpCount);

    UErrorCode status = U_ZERO_ERROR;
    u_strToUTF32(out.data(), static_cast<int32_t>(out.size()), nullptr, data, dataLen, &status);
    THROW_HR_IF(E_UNEXPECTED, status > U_ZERO_ERROR);

    return out;
}

static void foldStringUtf32(std::vector<UChar32>& str)
{
    for (auto& cp : str)
    {
        cp = u_foldCase(cp, U_FOLD_CASE_DEFAULT);
    }
}

static size_t trySkip(const std::vector<UChar32>& input, const UChar32 searchChar, size_t startIndex)
{
    for (size_t i = startIndex; i < input.size(); ++i)
    {
        if (input[i] == searchChar)
        {
            return i;
        }
    }
    return npos;
}

// Unlike the equivalent in fzf, this one does more than Unicode.
static size_t asciiFuzzyIndex(const std::vector<UChar32>& input, const std::vector<UChar32>& pattern)
{
    size_t idx = 0;
    size_t firstIdx = 0;
    for (size_t pi = 0; pi < pattern.size(); ++pi)
    {
        idx = trySkip(input, pattern[pi], idx);
        if (idx == npos)
        {
            return npos;
        }

        if (pi == 0 && idx > 0)
        {
            firstIdx = idx - 1;
        }

        idx++;
    }
    return firstIdx;
}

static int16_t calculateBonus(CharClass prevClass, CharClass currentClass)
{
    if (prevClass == CharClass::NonWord && currentClass != CharClass::NonWord)
    {
        return BoundaryBonus;
    }
    if ((prevClass == CharClass::CharLower && currentClass == CharClass::CharUpper) ||
        (prevClass != CharClass::Digit && currentClass == CharClass::Digit))
    {
        return CamelCaseBonus;
    }
    if (currentClass == CharClass::NonWord)
    {
        return NonWordBonus;
    }
    return 0;
}

static constexpr auto s_charClassLut = []() {
    std::array<CharClass, U_CHAR_CATEGORY_COUNT> lut{};
    lut.fill(CharClass::NonWord);
    lut[U_UPPERCASE_LETTER] = CharClass::CharUpper;
    lut[U_LOWERCASE_LETTER] = CharClass::CharLower;
    lut[U_MODIFIER_LETTER] = CharClass::CharLower;
    lut[U_OTHER_LETTER] = CharClass::CharLower;
    lut[U_DECIMAL_DIGIT_NUMBER] = CharClass::Digit;
    return lut;
}();

static CharClass classOf(UChar32 ch)
{
    return s_charClassLut[u_charType(ch)];
}

static int32_t fzfFuzzyMatchV2(const std::vector<UChar32>& text, const std::vector<UChar32>& pattern, std::vector<size_t>* pos)
{
    if (pattern.size() == 0)
    {
        return 0;
    }

    auto foldedText = text;
    foldStringUtf32(foldedText);

    size_t firstIndexOf = asciiFuzzyIndex(foldedText, pattern);
    if (firstIndexOf == npos)
    {
        return 0;
    }

    auto initialScores = std::vector<int16_t>(text.size());
    auto consecutiveScores = std::vector<int16_t>(text.size());
    auto firstOccurrenceOfEachChar = std::vector<size_t>(pattern.size());
    auto bonusesSpan = std::vector<int16_t>(text.size());

    int16_t maxScore = 0;
    size_t maxScorePos = 0;
    size_t patternIndex = 0;
    size_t lastIndex = 0;
    UChar32 firstPatternChar = pattern[0];
    UChar32 currentPatternChar = pattern[0];
    int16_t previousInitialScore = 0;
    CharClass previousClass = CharClass::NonWord;
    bool inGap = false;

    std::span<const UChar32> lowerText(foldedText);
    auto lowerTextSlice = lowerText.subspan(firstIndexOf);
    auto initialScoresSlice = std::span(initialScores).subspan(firstIndexOf);
    auto consecutiveScoresSlice = std::span(consecutiveScores).subspan(firstIndexOf);
    auto bonusesSlice = std::span(bonusesSpan).subspan(firstIndexOf, text.size() - firstIndexOf);

    for (size_t i = 0; i < lowerTextSlice.size(); i++)
    {
        const auto currentChar = lowerTextSlice[i];
        const auto currentClass = classOf(text[i + firstIndexOf]);
        const auto bonus = calculateBonus(previousClass, currentClass);
        bonusesSlice[i] = bonus;
        previousClass = currentClass;

        //currentPatternChar was already folded in ParsePattern
        if (currentChar == currentPatternChar)
        {
            if (patternIndex < pattern.size())
            {
                firstOccurrenceOfEachChar[patternIndex] = firstIndexOf + i;
                patternIndex++;
                if (patternIndex < pattern.size())
                {
                    currentPatternChar = pattern[patternIndex];
                }
            }
            lastIndex = firstIndexOf + i;
        }
        if (currentChar == firstPatternChar)
        {
            int16_t score = ScoreMatch + bonus * BonusFirstCharMultiplier;
            initialScoresSlice[i] = score;
            consecutiveScoresSlice[i] = 1;
            if (pattern.size() == 1 && (score > maxScore))
            {
                maxScore = score;
                maxScorePos = firstIndexOf + i;
                if (bonus == BoundaryBonus)
                {
                    break;
                }
            }
            inGap = false;
        }
        else
        {
            initialScoresSlice[i] = std::max<int16_t>(previousInitialScore + (inGap ? ScoreGapExtension : ScoreGapStart), 0);
            consecutiveScoresSlice[i] = 0;
            inGap = true;
        }
        previousInitialScore = initialScoresSlice[i];
    }

    if (patternIndex != pattern.size())
    {
        return 0;
    }

    if (pattern.size() == 1)
    {
        if (pos)
        {
            pos->push_back(maxScorePos);
        }
        return maxScore;
    }

    const auto firstOccurrenceOfFirstChar = firstOccurrenceOfEachChar[0];
    const auto width = lastIndex - firstOccurrenceOfFirstChar + 1;
    const auto rows = pattern.size();
    auto consecutiveCharMatrixSize = width * pattern.size();

    std::vector<int16_t> scoreMatrix(width * rows);
    std::copy_n(initialScores.begin() + firstOccurrenceOfFirstChar, width, scoreMatrix.begin());
    std::span scoreSpan(scoreMatrix);

    std::vector<int16_t> consecutiveCharMatrix(width * rows);
    std::copy_n(consecutiveScores.begin() + firstOccurrenceOfFirstChar, width, consecutiveCharMatrix.begin());
    std::span consecutiveCharMatrixSpan(consecutiveCharMatrix);

    auto patternSliceStr = std::span(pattern).subspan(1);

    for (size_t off = 0; off < pattern.size() - 1; off++)
    {
        auto patternCharOffset = firstOccurrenceOfEachChar[off + 1];
        auto sliceLen = lastIndex - patternCharOffset + 1;
        currentPatternChar = patternSliceStr[off];
        patternIndex = off + 1;
        auto row = patternIndex * width;
        inGap = false;
        std::span<const UChar32> textSlice = lowerText.subspan(patternCharOffset, sliceLen);
        std::span bonusSlice(bonusesSpan.begin() + patternCharOffset, textSlice.size());
        std::span<int16_t> consecutiveCharMatrixSlice = consecutiveCharMatrixSpan.subspan(row + patternCharOffset - firstOccurrenceOfFirstChar, textSlice.size());
        std::span<int16_t> consecutiveCharMatrixDiagonalSlice = consecutiveCharMatrixSpan.subspan(row + patternCharOffset - firstOccurrenceOfFirstChar - 1 - width, textSlice.size());
        std::span<int16_t> scoreMatrixSlice = scoreSpan.subspan(row + patternCharOffset - firstOccurrenceOfFirstChar, textSlice.size());
        std::span<int16_t> scoreMatrixDiagonalSlice = scoreSpan.subspan(row + patternCharOffset - firstOccurrenceOfFirstChar - 1 - width, textSlice.size());
        std::span<int16_t> scoreMatrixLeftSlice = scoreSpan.subspan(row + patternCharOffset - firstOccurrenceOfFirstChar - 1, textSlice.size());

        if (!scoreMatrixLeftSlice.empty())
        {
            scoreMatrixLeftSlice[0] = 0;
        }

        for (size_t j = 0; j < textSlice.size(); j++)
        {
            const auto currentChar = textSlice[j];
            const auto column = patternCharOffset + j;
            const int16_t score = inGap ? scoreMatrixLeftSlice[j] + ScoreGapExtension : scoreMatrixLeftSlice[j] + ScoreGapStart;
            int16_t diagonalScore = 0;
            int16_t consecutive = 0;
            if (currentChar == currentPatternChar)
            {
                diagonalScore = scoreMatrixDiagonalSlice[j] + ScoreMatch;
                int16_t bonus = bonusSlice[j];
                consecutive = consecutiveCharMatrixDiagonalSlice[j] + 1;
                if (bonus == BoundaryBonus)
                {
                    consecutive = 1;
                }
                else if (consecutive > 1)
                {
                    bonus = std::max({ bonus, BonusConsecutive, (bonusesSpan[column - consecutive + 1]) });
                }
                if (diagonalScore + bonus < score)
                {
                    diagonalScore += bonusSlice[j];
                    consecutive = 0;
                }
                else
                {
                    diagonalScore += bonus;
                }
            }
            consecutiveCharMatrixSlice[j] = consecutive;
            inGap = (diagonalScore < score);
            int16_t cellScore = std::max(int16_t{ 0 }, std::max(diagonalScore, score));
            if (off + 2 == pattern.size() && cellScore > maxScore)
            {
                maxScore = cellScore;
                maxScorePos = column;
            }
            scoreMatrixSlice[j] = cellScore;
        }
    }

    size_t currentColIndex = maxScorePos;
    if (pos)
    {
        patternIndex = pattern.size() - 1;
        bool preferCurrentMatch = true;
        while (true)
        {
            const auto rowStartIndex = patternIndex * width;
            const auto colOffset = currentColIndex - firstOccurrenceOfFirstChar;
            const auto cellScore = scoreMatrix[rowStartIndex + colOffset];
            int32_t diagonalCellScore = 0;
            int32_t leftCellScore = 0;

            if (patternIndex > 0 && currentColIndex >= firstOccurrenceOfEachChar[patternIndex])
            {
                diagonalCellScore = scoreMatrix[rowStartIndex - width + colOffset - 1];
            }
            if (currentColIndex > firstOccurrenceOfEachChar[patternIndex])
            {
                leftCellScore = scoreMatrix[rowStartIndex + colOffset - 1];
            }

            if (cellScore > diagonalCellScore &&
                (cellScore > leftCellScore || (cellScore == leftCellScore && preferCurrentMatch)))
            {
                pos->push_back(currentColIndex);
                if (patternIndex == 0)
                {
                    break;
                }
                patternIndex--;
            }

            currentColIndex--;
            if (rowStartIndex + colOffset >= consecutiveCharMatrixSize)
            {
                break;
            }

            preferCurrentMatch = (consecutiveCharMatrix[rowStartIndex + colOffset] > 1) ||
                                 ((rowStartIndex + width + colOffset + 1 <
                                   consecutiveCharMatrixSize) &&
                                  (consecutiveCharMatrix[rowStartIndex + width + colOffset + 1] > 0));
        }
    }
    return maxScore;
}

Pattern fzf::matcher::ParsePattern(const std::wstring_view patternStr)
{
    Pattern patObj;
    size_t pos = 0;

    while (true)
    {
        const auto beg = patternStr.find_first_not_of(L' ', pos);
        if (beg == std::wstring_view::npos)
        {
            break; // No more non-space characters
        }

        const auto end = std::min(patternStr.size(), patternStr.find_first_of(L' ', beg));
        const auto word = patternStr.substr(beg, end - beg);
        auto codePoints = utf16ToUtf32(word);
        foldStringUtf32(codePoints);
        patObj.terms.push_back(std::move(codePoints));
        pos = end;
    }

    return patObj;
}

std::optional<MatchResult> fzf::matcher::Match(std::wstring_view text, const Pattern& pattern)
{
    if (pattern.terms.empty())
    {
        return MatchResult{};
    }

    const auto textCodePoints = utf16ToUtf32(text);

    int32_t totalScore = 0;
    std::vector<size_t> allUtf32Pos;

    for (const auto& term : pattern.terms)
    {
        std::vector<size_t> termPos;
        auto score = fzfFuzzyMatchV2(textCodePoints, term, &termPos);
        if (score <= 0)
        {
            return std::nullopt;
        }

        totalScore += score;
        allUtf32Pos.insert(allUtf32Pos.end(), termPos.begin(), termPos.end());
    }

    std::ranges::sort(allUtf32Pos);
    allUtf32Pos.erase(std::ranges::unique(allUtf32Pos).begin(), allUtf32Pos.end());

    std::vector<TextRun> runs;
    std::size_t nextCodePointPos = 0;
    size_t utf16Offset = 0;

    bool inRun = false;
    size_t runStart = 0;

    for (size_t cpIndex = 0; cpIndex < textCodePoints.size(); cpIndex++)
    {
        const auto cp = textCodePoints[cpIndex];
        const size_t cpWidth = U16_LENGTH(cp);

        const bool isMatch = (nextCodePointPos < allUtf32Pos.size() && allUtf32Pos[nextCodePointPos] == cpIndex);
        if (isMatch)
        {
            if (!inRun)
            {
                runStart = utf16Offset;
                inRun = true;
            }
            nextCodePointPos++;
        }
        else if (inRun)
        {
            runs.push_back({ runStart, utf16Offset - 1 });
            inRun = false;
        }

        utf16Offset += cpWidth;
    }

    if (inRun)
    {
        runs.push_back({ runStart, utf16Offset - 1 });
    }

    return MatchResult{ totalScore, std::move(runs) };
}

File: rust/terminal-app/src/fzf.rs
const SCORE_MATCH: i32 = 16;
const SCORE_GAP_START: i32 = -3;
const SCORE_GAP_EXTENSION: i32 = -1;
const BOUNDARY_BONUS: i32 = SCORE_MATCH / 2;
const NON_WORD_BONUS: i32 = SCORE_MATCH / 2;
const CAMEL_CASE_BONUS: i32 = BOUNDARY_BONUS + SCORE_GAP_EXTENSION;
const BONUS_CONSECUTIVE: i32 = -(SCORE_GAP_START + SCORE_GAP_EXTENSION);
const BONUS_FIRST_CHAR_MULTIPLIER: i32 = 2;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum CharClass {
    NonWord,
    Lower,
    Upper,
    Digit,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Pattern {
    terms: Vec<Vec<char>>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct TextRun {
    pub start: usize,
    pub end: usize,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct MatchResult {
    pub score: i32,
    pub runs: Vec<TextRun>,
}

#[must_use]
pub fn parse_pattern(pattern: &str) -> Pattern {
    let terms = pattern
        .split(' ')
        .filter(|term| !term.is_empty())
        .map(|term| term.chars().map(simple_fold).collect())
        .collect();

    Pattern { terms }
}

#[must_use]
pub fn match_text(text: &str, pattern: &Pattern) -> Option<MatchResult> {
    if pattern.terms.is_empty() {
        return Some(MatchResult::default());
    }

    let text_code_points = text.chars().collect::<Vec<_>>();
    let mut total_score = 0;
    let mut all_positions = Vec::new();

    for term in &pattern.terms {
        let mut term_positions = Vec::new();
        let score = fuzzy_match_v2(&text_code_points, term, &mut term_positions);
        if score <= 0 {
            return None;
        }

        total_score += score;
        all_positions.extend(term_positions);
    }

    all_positions.sort_unstable();
    all_positions.dedup();

    let mut runs = Vec::new();
    let mut next_position = 0;
    let mut utf16_offset = 0;
    let mut run_start = None;

    for (code_point_index, code_point) in text_code_points.iter().copied().enumerate() {
        let code_point_width = code_point.len_utf16();
        let is_match = all_positions.get(next_position).copied() == Some(code_point_index);

        if is_match {
            run_start.get_or_insert(utf16_offset);
            next_position += 1;
        } else if let Some(start) = run_start.take() {
            runs.push(TextRun {
                start,
                end: utf16_offset - 1,
            });
        }

        utf16_offset += code_point_width;
    }

    if let Some(start) = run_start {
        runs.push(TextRun {
            start,
            end: utf16_offset - 1,
        });
    }

    Some(MatchResult {
        score: total_score,
        runs,
    })
}

fn simple_fold(code_point: char) -> char {
    match code_point {
        '\u{03c2}' => '\u{03c3}',
        _ => code_point.to_lowercase().next().unwrap_or(code_point),
    }
}

fn class_of(code_point: char) -> CharClass {
    if code_point.is_uppercase() {
        CharClass::Upper
    } else if code_point.is_lowercase() || code_point.is_alphabetic() {
        CharClass::Lower
    } else if code_point.is_numeric() {
        CharClass::Digit
    } else {
        CharClass::NonWord
    }
}

fn calculate_bonus(previous: CharClass, current: CharClass) -> i32 {
    if previous == CharClass::NonWord && current != CharClass::NonWord {
        BOUNDARY_BONUS
    } else if (previous == CharClass::Lower && current == CharClass::Upper)
        || (previous != CharClass::Digit && current == CharClass::Digit)
    {
        CAMEL_CASE_BONUS
    } else if current == CharClass::NonWord {
        NON_WORD_BONUS
    } else {
        0
    }
}

fn try_skip(input: &[char], search: char, start: usize) -> Option<usize> {
    input[start..]
        .iter()
        .position(|candidate| *candidate == search)
        .map(|offset| start + offset)
}

fn fuzzy_index(input: &[char], pattern: &[char]) -> Option<usize> {
    let mut index = 0;
    let mut first_index = 0;

    for (pattern_index, search) in pattern.iter().copied().enumerate() {
        index = try_skip(input, search, index)?;
        if pattern_index == 0 && index > 0 {
            first_index = index - 1;
        }
        index += 1;
    }

    Some(first_index)
}

#[allow(clippy::too_many_lines)]
fn fuzzy_match_v2(text: &[char], pattern: &[char], positions: &mut Vec<usize>) -> i32 {
    if pattern.is_empty() {
        return 0;
    }

    let folded_text = text.iter().copied().map(simple_fold).collect::<Vec<_>>();
    let Some(first_index_of) = fuzzy_index(&folded_text, pattern) else {
        return 0;
    };

    let mut initial_scores = vec![0_i32; text.len()];
    let mut consecutive_scores = vec![0_usize; text.len()];
    let mut first_occurrence = vec![0_usize; pattern.len()];
    let mut bonuses = vec![0_i32; text.len()];

    let mut max_score = 0;
    let mut max_score_position = 0;
    let mut pattern_index = 0;
    let mut last_index = 0;
    let first_pattern_char = pattern[0];
    let mut current_pattern_char = pattern[0];
    let mut previous_initial_score = 0;
    let mut previous_class = CharClass::NonWord;
    let mut in_gap = false;

    for absolute_index in first_index_of..folded_text.len() {
        let current_char = folded_text[absolute_index];
        let current_class = class_of(text[absolute_index]);
        let bonus = calculate_bonus(previous_class, current_class);
        bonuses[absolute_index] = bonus;
        previous_class = current_class;

        if current_char == current_pattern_char {
            if pattern_index < pattern.len() {
                first_occurrence[pattern_index] = absolute_index;
                pattern_index += 1;
                if pattern_index < pattern.len() {
                    current_pattern_char = pattern[pattern_index];
                }
            }
            last_index = absolute_index;
        }

        if current_char == first_pattern_char {
            let score = SCORE_MATCH + bonus * BONUS_FIRST_CHAR_MULTIPLIER;
            initial_scores[absolute_index] = score;
            consecutive_scores[absolute_index] = 1;

            if pattern.len() == 1 && score > max_score {
                max_score = score;
                max_score_position = absolute_index;
                if bonus == BOUNDARY_BONUS {
                    break;
                }
            }
            in_gap = false;
        } else {
            initial_scores[absolute_index] = (previous_initial_score
                + if in_gap {
                    SCORE_GAP_EXTENSION
                } else {
                    SCORE_GAP_START
                })
            .max(0);
            consecutive_scores[absolute_index] = 0;
            in_gap = true;
        }

        previous_initial_score = initial_scores[absolute_index];
    }

    if pattern_index != pattern.len() {
        return 0;
    }

    if pattern.len() == 1 {
        positions.push(max_score_position);
        return max_score;
    }

    let first_occurrence_of_first_char = first_occurrence[0];
    let width = last_index - first_occurrence_of_first_char + 1;
    let rows = pattern.len();
    let matrix_size = width * rows;

    let mut score_matrix = vec![0_i32; matrix_size];
    score_matrix[..width].copy_from_slice(
        &initial_scores[first_occurrence_of_first_char..first_occurrence_of_first_char + width],
    );

    let mut consecutive_matrix = vec![0_usize; matrix_size];
    consecutive_matrix[..width].copy_from_slice(
        &consecutive_scores[first_occurrence_of_first_char..first_occurrence_of_first_char + width],
    );

    for offset in 0..pattern.len() - 1 {
        let pattern_char_offset = first_occurrence[offset + 1];
        let slice_len = last_index - pattern_char_offset + 1;
        current_pattern_char = pattern[offset + 1];
        pattern_index = offset + 1;
        let row = pattern_index * width;
        in_gap = false;
        let start_column = pattern_char_offset - first_occurrence_of_first_char;

        if start_column > 0 {
            score_matrix[row + start_column - 1] = 0;
        }

        for relative_index in 0..slice_len {
            let column = pattern_char_offset + relative_index;
            let column_offset = column - first_occurrence_of_first_char;
            let left_score = if relative_index == 0 {
                0
            } else {
                score_matrix[row + column_offset - 1]
            };
            let score = left_score
                + if in_gap {
                    SCORE_GAP_EXTENSION
                } else {
                    SCORE_GAP_START
                };

            let mut diagonal_score = 0;
            let mut consecutive = 0;

            if folded_text[column] == current_pattern_char {
                diagonal_score = score_matrix[row - width + column_offset - 1] + SCORE_MATCH;
                let mut bonus = bonuses[column];
                consecutive = consecutive_matrix[row - width + column_offset - 1] + 1;

                if bonus == BOUNDARY_BONUS {
                    consecutive = 1;
                } else if consecutive > 1 {
                    let chain_start = column + 1 - consecutive;
                    bonus = bonus.max(BONUS_CONSECUTIVE).max(bonuses[chain_start]);
                }

                if diagonal_score + bonus < score {
                    diagonal_score += bonuses[column];
                    consecutive = 0;
                } else {
                    diagonal_score += bonus;
                }
            }

            consecutive_matrix[row + column_offset] = consecutive;
            in_gap = diagonal_score < score;
            let cell_score = 0.max(diagonal_score.max(score));

            if offset + 2 == pattern.len() && cell_score > max_score {
                max_score = cell_score;
                max_score_position = column;
            }

            score_matrix[row + column_offset] = cell_score;
        }
    }

    let mut current_column = max_score_position;
    pattern_index = pattern.len() - 1;
    let mut prefer_current_match = true;

    loop {
        let row_start = pattern_index * width;
        let column_offset = current_column - first_occurrence_of_first_char;
        let cell_score = score_matrix[row_start + column_offset];
        let diagonal_score =
            if pattern_index > 0 && current_column >= first_occurrence[pattern_index] {
                score_matrix[row_start - width + column_offset - 1]
            } else {
                0
            };
        let left_score = if current_column > first_occurrence[pattern_index] {
            score_matrix[row_start + column_offset - 1]
        } else {
            0
        };

        if cell_score > diagonal_score
            && (cell_score > left_score || (cell_score == left_score && prefer_current_match))
        {
            positions.push(current_column);
            if pattern_index == 0 {
                break;
            }
            pattern_index -= 1;
        }

        if current_column == 0 {
            break;
        }
        current_column -= 1;

        if row_start + column_offset >= matrix_size {
            break;
        }

        prefer_current_match = consecutive_matrix[row_start + column_offset] > 1
            || (row_start + width + column_offset + 1 < matrix_size
                && consecutive_matrix[row_start + width + column_offset + 1] > 0);
    }

    max_score
}
